File: 2025/R/solve_day3_part2.R
solve_day3_part2 <- function(input) {
  input |>
    strsplit("") |>
    lapply(as.integer) |>
    vapply(max_jolt, numeric(1)) |>
    sum()
}

#' @params bank Character vector of digits
#' @params digit_count Numeric number of digits to use for joltage
max_jolt <- function(bank, digit_count = 12) {
  digit_nums <- seq_len(digit_count)
  digits <- character(digit_count)
  prev_loc <- 0L
  len <- length(bank)

  for (i in digit_nums) {
    max_loc <- len - (digit_count - i)
    valid_bank <- bank[(prev_loc + 1L):max_loc]
    loc <- which.max(valid_bank)
    prev_loc <- prev_loc + loc
    digits[i] <- valid_bank[loc]
  }

  joltage <- paste0(digits, collapse = "")

  as.numeric(joltage)
}

File: python/solve_day3_part2.py
def solve_day3_part2(text):
    banks = [list(line.strip("\n")) for line in text]
    joltage = [max_jolt(bank) for bank in banks]

    return(sum(joltage))

def max_jolt(bank, n = 12):
    digits = [None] * n
    prev_loc = -1
    bank_len = len(bank)

    for i in range(n):
        max_loc = bank_len - (n - i) + 1

        search_bank = bank[(prev_loc + 1):max_loc]
        digits[i] = max(search_bank)

        loc = search_bank.index(digits[i])
        prev_loc = prev_loc + loc + 1

    result = int(''.join(digits))

    return(result)
